### ISSUE TITLE
print(ma.array) neglects print precision

### ISSUE THREAD

**Issue by redkuul** [numpy/numpy#17236]
The default print behavior of ma is not what I expected. 
I expected that when precision is set, this would also be reflected in print(mx)
This seems to work as expected with print(mx.data) or print(mx.filled()) 

```python
import numpy as np
import numpy.ma as ma
mx=ma.array(data=[1.12345,2.12345,3.12345,-1.12345,5.12345],mask=[False, False, False, True, False], fill_value=np.nan)
np.set_printoptions(precision=2,floatmode='fixed')

print(mx) #expected output: [1.12, 2.12, 3.12,--,5.12] -> but I get: [1.12345,2.12345,3.12345,--,5.12345]
print(mx.data) # expected output [1.12, 2.12, 3.12,-1.12,5.12] -> OK
print(mx.filled()) # expected output [1.12, 2.12, 3.12,nan,5.12] -> OK
```

### Numpy/Python version information:
1.18.5 3.8.3 (default, Jul  2 2020, 17:30:36) [MSC v.1916 64 bit (AMD64)]

**Comment by eric-wieser** [comment]
Yep, this is because printing of masked arrays works by first converting to an object array, then inserting the `--`. I had a patch to rework the printing completely a long time ago, but forget what happened to it.

This would definitely nice to fix, but I suspect it's rather tricky.

**Comment by mikeiovine** [comment]
I found a hacky workaround for this. For a masked array `x`, `np.isfinite(x)` returns `False` wherever `x.mask = True`. So, we can avoid converting the array to `object` by doing something like the following:

In `MaskedArray`: 
```python
def __str__(self):
    return np.array2string(self) 
```
In the formatters in `arrayprint`, eg `FloatingFormat`
```python
def __call__(self, x):
    if not np.isfinite(x):
        with errstate(invalid='ignore'):
            if np.isnan(x):
                # insert nan
            elif np.isinf(x):
                # insert inf.
            else:
                # insert the "--"
```
I hesitate to do this because I'm not sure if it's the intended behavior of `np.isfinite`; I don't want to rely on it if it's a bug. But if this change is fine, I can come up with a fix and submit a pull request.

**Comment by athompson1991** [comment]
Hi @eric-wieser and @redkuul. I was sent this issue through the [Code Triage](https://www.codetriage.com/) app and have put together a branch that has a new test and bug fix:

https://github.com/athompson1991/numpy/tree/issue-17236

All the tests are passing (not counting four that seem to be failing even on up-to-date master branch), and the `ma.array` / `ma.MaskedArray` type is printing correctly now. However I am not sure the fix is ideal, as it's doing a sort of ham-handed string manipulation in the `__str__` method. Let me know what you think though, I see how the printing is tricky but this does indeed fix the immediate bug @redkuul describes.

It's worth noting that this is the first time I've ever tried contributing anything to an open source project, so I am appreciative of any feedback and am trying to follow best practices to best of ability as per documentation.

**Comment by mattip** [comment]
Welcome @athompson1991. It seems that branch could be [turned into a PR](https://github.com/numpy/numpy/compare/master...athompson1991:issue-17236). I see it has a test and some code changes. After a quick lookover, I think your `is_float` check can be done differently. Probably the best way to get more eyes on it is to turn it into a PR.

> All the tests are passing (not counting four that seem to be failing even on up-to-date master branch)

Hmm, that is worrisome, but maybe your build is a little different that the CI, which is why making it a PR might be good.
